Add support for more metadata in UVCal, add metadata_only support

bhazelton opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
Add support for telescope_location, antenna_positions and lst_array in UVCal object and file types. Add support for a metadata_only mode in UVCal, including options to only read the metadata when reading in calibration files.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. If this PR closes an issue, put the word 'closes' before the issue link to auto-close the issue when the PR is merged. -->
This is preparing for creating a function to initialize a UVCal object from a UVData object (#753)
closes #735 

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation change (documentation changes only)
- [ ] Version change
- [ ] Build or continuous integration change


## Checklist:
<!--- You may remove the checklists that don't apply to your change type(s) or just leave them empty -->
<!--- Go over all the following points, and replace the space with an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the [contribution guide](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/master/.github/CONTRIBUTING.md).
- [x] My code follows the code style of this project.

New feature checklist:
- [x] I have added or updated the docstrings associated with my feature using the [numpy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html).
- [ ] I have updated the tutorial to highlight my new feature (if appropriate).
- [x] I have added tests to cover my new feature.
- [x] All new and existing tests pass.
- [ ] I have updated the [CHANGELOG](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/master/CHANGELOG.md).

Breaking change checklist:
- [x] I have updated the docstrings associated with my change using the [numpy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html).
- [ ] I have updated the tutorial to reflect my changes (if appropriate).
- [x] My change includes backwards compatibility and deprecation warnings (if possible).
- [x] I have added tests to cover my changes.
- [x] All new and existing tests pass.
- [ ] I have updated the [CHANGELOG](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/master/CHANGELOG.md).

dstorer reviewed (approved):
This looks good to me - thanks for implementing this!

bhazelton commented:
@mkolopanis I'm afraid this PR adds more deprecation warnings which is upsetting hera_qm tests.

bhazelton commented:
@jsdillon This PR is starting a process to begin requiring some additional metadata on UVCal objects: telescope_location, lst_array, and antenna_positions. There have been many requests for these to exist on UVCal objects from users, mostly HERA users.

I'm working on a function to transfer metadata (including those) from a UVData object, but wanted to give you an early heads up about this change. If you're concerned about it in any way please let us know!

jsdillon commented:
Thanks for the heads up @bhazelton. Are these extra parameters going to be required on read, or just on write? We've got a lot of calibration solution files sitting around that don't have these and it would be a real pain to have to update them all or rewrite code that reads them. 

I'd be happy if, on read, missing parameters produces a warning, and on write, they produce an error.

bhazelton commented:
@jsdillon the telescope_location and lst_array will be totally fine because as long as the telescope_name is in our known_telescopes (and HERA is), those can be set by the reader on the object.

The antenna_positions are trickier. I was thinking we would want to make that a required parameter eventually (although the time scale is negotiable) in which case it would error if you run a check on the object. We can always make it not error on the read by setting the `run_check` keyword to False, but then none of the checks are run. I suppose we could add another option that was specific to that parameter. Or we could decide to only make that parameter optional or put transitioning it to required on a longer time scale than the others. What do you think?

jsdillon commented:
For telescope_location, the one issue I might imagine is with the RIMEz simulations, for example those used in HERA validation. Perhaps @zacharymartinot can work with you to include RIMEz in pyuvdata appropriately?

For the antenna locations, this is tricky. I'd love for all UVCal objects to have it, but backwards compatibility in pipelines is a concern (and I'd be hesitant to turn off checking in any pipeline steps). Is it possible to have a parameter that is required for writing but not for reading? Otherwise, I suspect we'll want a long deprecation period.

bhazelton commented:
@jsdillon there is not support currently for something that is required on write but not on read in the structure of pyuvdata objects. It's a bit strange to have something like that because then you could read it in but not write it back out without adding the antenna_positions. Is that really a useful mechanism?

jsdillon commented:
If it existed, it might have helped, but I don't think it's worth building. 

A long warning period is probably necessary.

bhazelton commented:
@jsdillon We came up with a better solution in our meeting this morning. We will add antenna positions as optional metadata to our known telescopes list. I will add the HERA positions and then if they are missing we can add them to the object using that list. That will make sure they are defined on the object (at least for HERA data) even if they are missing in the file that is read in.

jsdillon commented:
Good idea! That sounds fine to me.

theo-kunicki reviewed (approved):
I've looked over this, and it seems like these changes will be very useful -- thank you, Bryna! I'll let you know if I run into any problems once it's been merged, but I can't see anything obvious right now.

plaplant reviewed (commented):
Thanks so much @bhazelton! This is a tremendous amount of work, and the overhaul of UVCal tests was long-overdue. I have a few minor comments, but generally everything looks excellent.

plaplant commented:
Also we definitely want to make sure we update the changelog.

bhazelton commented on pyuvdata/uvcal/calfits.py:
```diff
@@ -126,6 +126,13 @@ def write_calfits(
+            prihdr["ARRAYX"] = self.telescope_location[0]
+            prihdr["ARRAYY"] = self.telescope_location[1]
+            prihdr["ARRAYZ"] = self.telescope_location[2]
+            prihdr["LAT     "] = self.telescope_location_lat_lon_alt_degrees[0]
```
I think it doesn't matter and I'm happy to change it. The FITS format forces them all to be length 8 and adds padding if they're not that long.

bhazelton commented on pyuvdata/uvdata/fhd.py:
```diff
@@ -343,174 +561,37 @@ def read_fhd(
-                    value = value.decode("utf8")
-
-                self.extra_keywords[keyword.upper()] = value
+            (
```
Yeah, that's a good idea. I'll make that change.

bhazelton commented on pyuvdata/uvcal/fhd_cal.py:
```diff
@@ -62,159 +74,126 @@ def read_fhd_cal(
+        # (e.g. 1.999426... rather than 2)
+        time_res = obs_data["TIME_RES"][0]
+        # time_res is constrained to be a scalar currently
+        self.integration_time = np.float(time_res)
```
Sure, happy to change it.

bhazelton commented on pyuvdata/uvcal/fhd_cal.py:
```diff
@@ -62,159 +74,126 @@ def read_fhd_cal(
-        self.freq_array[0, :] = cal_data["freq"][0]
-        self.channel_width = float(np.mean(np.diff(self.freq_array)))
+        self.Nfreqs = int(obs_data["N_FREQ"][0])
+        self.freq_array = np.zeros((1, len(bl_info["FREQ"][0])), dtype=np.float_)
```
Sure, happy to change it.

bhazelton commented:
Thanks for the review @plaplant!! I know it was a lot of changes. I think I addressed all your comments and I updated the changelog. Can you take a look at the changelog and make sure it all makes sense? There were a lot of changes.

plaplant reviewed (approved):
Thanks for making the changes @bhazelton. I think the changelog updates make sense. Feel free to merge anytime. We should also update the hera_qm tests after this has been merged.
